csharp/ql/src/semmlecode.csharp.dbscheme

```diff
@@ -479,6 +479,34 @@ type_mention_location(
   unique int id: @type_mention ref,
   int loc: @location ref);
 
+@has_type_annotation = @assignable | @type_parameter | @callable | @array_type | @expr | @delegate_type;
+
+/**
+ * A direct annotation on an entity, for example `string? x;`.
+ *
+ * Annotations:
+ * 0 = reftype nullability is not applicable
```

Comment:
How is 0 different from no type annotation?

Reply:
Same question for 1 and 2.

Reply:
It gives the *reason* why there is no nullable annotation. It could be because

0) The type cannot have a nullability
1) The type wasn't compiled in a nullable context. Either before C#8, OR, in C#8 but the nullable context was disabled
2) The code is compiled in a nullable context, but the nullable annotation was not specified.

This is good information. Without this information, you couldn't tell whether `Foo x` was nullable or not. You need to distinguish between `Foo` in a nullable context (`Foo!`), `Foo` in a non-nullable context (`Foo`), or `Foo?`.

You could however deduce case 0 by the absence of other nullability annotations since they are mutually exclusive.

Reply:
> You need to distinguish between Foo in a nullable context (Foo!), Foo in a non-nullable context (Foo), or Foo?.

To me that sounds like two annotations; either `!` or `?`. Perhaps I would better understand with concrete examples of cases 0-3.

Reply:
The question you are asking, I think, is why does Roslyn's `NullableAnnotation` have 4 values, and not just a boolean "nullable or not". The reason is because we want to distinguish between legacy code and code written in a nullable context. In particular, nullness queries could behave differently, and we don't want to render types with a `?` when that's not how they were written.

In particular, could we conflate `string` and `string?` ? I think not. `string` may never be null, whereas `string?` has a real possibility of being null.

Consider
```
#nullable enable
string getName1();
string? getName2();

#nullable disable
string getName3();  
```
In this case, `getName1` returns a non-null string. `getName2` may very well return null, and `getName3` is unspecified.

Either way, I think it's good information to have in the database and there's no good reason to throw it away.